extensions/scarb-execute/src/oracle/jsonrpc.rs

```diff
@@ -0,0 +1,268 @@
+//! Simple JSON-RPC types that suit our needs and nothing more.
+//! It is surprising how complex existing JSON-RPC Rust implementations are,
+//! we don't need any kinds of async, extendibility or network stuff that they provide.
+
+#![allow(dead_code, unused)] // TODO: Remove in next PRs
+
+use anyhow::{Result, bail};
+use serde::de::{Unexpected, Visitor};
+use serde::{Deserialize, Deserializer, Serialize, Serializer, de};
+use std::fmt;
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(untagged)]
+pub enum Message {
+    Request(Request),
+    Response(Response),
+    Notification(Notification),
+}
+
+impl Message {
+    pub fn kind(&self) -> &'static str {
+        match self {
+            Message::Request(_) => "request",
+            Message::Response(_) => "response",
+            Message::Notification(_) => "notification",
+        }
+    }
+
+    pub fn expect_request(self) -> Result<Request> {
+        match self {
+            Message::Request(request) => Ok(request),
+            _ => bail!("expected request, got {}", self.kind()),
+        }
+    }
+
+    pub fn expect_response(self) -> Result<Response> {
+        match self {
+            Message::Response(response) => Ok(response),
+            _ => bail!("expected response, got {}", self.kind()),
+        }
+    }
+
+    pub fn expect_notification(self) -> Result<Notification> {
+        match self {
+            Message::Notification(notification) => Ok(notification),
+            _ => bail!("expected notification, got {}", self.kind()),
+        }
+    }
+}
+
+impl From<Request> for Message {
+    fn from(request: Request) -> Message {
+        Message::Request(request)
+    }
+}
+
+impl From<Response> for Message {
+    fn from(response: Response) -> Message {
+        Message::Response(response)
+    }
+}
+
+impl From<Notification> for Message {
+    fn from(notification: Notification) -> Message {
+        Message::Notification(notification)
+    }
+}
+
+impl TryFrom<Message> for Request {
+    type Error = anyhow::Error;
+
+    fn try_from(message: Message) -> Result<Self, Self::Error> {
+        message.expect_request()
+    }
+}
+
+impl TryFrom<Message> for Response {
+    type Error = anyhow::Error;
+
+    fn try_from(message: Message) -> Result<Self, Self::Error> {
+        message.expect_response()
+    }
+}
+
+impl TryFrom<Message> for Notification {
+    type Error = anyhow::Error;
+
+    fn try_from(message: Message) -> Result<Self, Self::Error> {
+        message.expect_notification()
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Request {
+    pub jsonrpc: TwoPointZero,
+    pub id: Id,
+    pub method: String,
+    #[serde(default = "serde_json::Value::default")]
+    #[serde(skip_serializing_if = "serde_json::Value::is_null")]
+    pub params: serde_json::Value,
```

Comment:
Should we ensure somewhere, that this is `Value::Object`, as spec says "A **Structured** value"?